### PR TITLE
Span<T> api update

### PR DESCRIPTION
**Pull request by mattwarren**
Changing method/property order to match CoreFX impl (to make diffs easier)

Added other missing methods (to match CoreFX impl):
- `public void CopyTo(Span<T> destination)`
- `public static bool operator ==(Span<T> left, Span<T> right)`
- `public static bool operator !=(Span<T> left, Span<T> right)`
- `public override bool Equals(object obj)`
- `public override int GetHashCode()`

Also removed `public void Set(ReadOnlySpan<T> values)` as it's no longer part of the [Span<T> public API](https://github.com/dotnet/apireviews/tree/master/2016/11-04-SpanOfT#spantset)

### CODE REVIEW

**Comment by mattwarren** [comment]
I wasn't able to add `[EditorBrowsable(EditorBrowsableState.Never)]` to `public override bool Equals(object obj)` and `public override int GetHashCode()`, if I do this:

```
using EditorBrowsableState = System.ComponentModel.EditorBrowsableState;
using EditorBrowsableAttribute = System.ComponentModel.EditorBrowsableAttribute;
```

I get the error:

```
the type or namespace name 'ComponentModel' does not exist in the namespace 'System'
```

Should I be able to do this?

**Comment by mattwarren** [comment]
One other thing, would you like me to change the simple properties to the nicer syntax? i.e. change this:

```
public int Length
{
    get { return _length; }
}
```
to `public int Length => _length;`

And then repeat this for the other properties where it's done like this in CoreFX?

File: src/mscorlib/src/System/Span.cs [hunk 1]
```diff
@@ -15,18 +15,13 @@ namespace System
     /// </summary>
     public struct Span<T>
     {
-        /// <summary>A byref or a native ptr.</summary>
```
Comment (jkotas) [inline src/mscorlib/src/System/Span.cs]: The convention is to keep fields at the top. (It was recently added to https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md to make it explicit.)

**Comment by jkotas** [comment]
> would you like me to change the simple properties to the nicer syntax

Yes - we do change the simple properties to the nicer syntax as we touch the code.

**Comment by jkotas** [comment]
> EditorBrowsableState 

There is discussion about this one in context of other changes: https://github.com/dotnet/corefx/issues/13746#issuecomment-266176583 . I would ignore it for this round of Span updates, until the discussion is settled.

**Comment by jkotas** [comment]
LGTM otherwise. Thanks for cleaning this up!

**Comment by jkotas** [comment]
cc @ahsonkhan @KrzysztofCwalina

**Comment by mattwarren** [comment]
@jkotas Mostly for my own interest, if you have code like this:

```
static void TestMethod()
{
  try {
    var data = new char[50];
    var slice = new Span<char>(data);
    Console.WriteLine("Before cast to object");
    Object sliceObj = (object)slice;
  }
  catch (Exception ex) {
    Console.WriteLine(ex.ToString());
  }
}
```

it currently throws `System.InvalidProgramException: Common Language Runtime detected an invalid program.`, because of the cast from `Span<T>` to `object. It seemingly happens at the point the method is JITted, because 'Before cast to object' is not printed.

In the future is the plan for the compiler to catch this, the CLR/Runtime or both? Also is it related to #8516?

**Comment by jkotas** [comment]
> In the future is the plan for the compiler to catch this, the CLR/Runtime or both?

Yes, the plan is for the C# compiler to catch this as well.

**Comment by jkotas** [comment]
Thank you!

**Comment by KrzysztofCwalina** [comment]
Yeah. Thanks for the cleanup!

**Comment by karelz** [comment]
Here's issue tracking decision on the EditorBrowsable attribute moving down to Private.CorLib - #14447.

**Pull request merged**